core/src/schema/content/number.rs

```diff
@@ -66,224 +66,266 @@ macro_rules! number_content {
-                        let inner = number_content::$as::deserialize(v).map_err(D::Error::custom)?;
-                        Ok(NumberContent::$as(inner))
-                    }
-                    else
-                )*
-                    {
-                        //TODO: generate static array with variant names and pass it to
-                        //      Error::unknown_variant()
-                        Err(D::Error::unknown_variant(format!("{:?}", subtype).as_str(), &[""]))
-                    }
+            fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+            where
+                S: Serializer,
+            {
+                match &self {
+                    $(
+                        NumberContent::$as(value) => {
+                            let mut obj = serde_json::to_value(value).map_err(S::Error::custom)?;
+                            let obj_map = obj.as_object_mut().ok_or(S::Error::custom("Object value expected"))?;
+                            obj_map.insert("subtype".to_string(), serde_convert_case!($as, "snake_case"));
+                            obj_map.serialize(serializer)
+                        }
+                    )*
                 }
-                None => {
-                    if let Ok(inner) = number_content::U64::deserialize(&v) {
-                        Ok(NumberContent::U64(inner))
-                    } else
-                    if let Ok(inner) = number_content::I64::deserialize(&v) {
-                        Ok(NumberContent::I64(inner))
-                    } else
-                    if let Ok(inner) = number_content::F64::deserialize(&v) {
-                        Ok(NumberContent::F64(inner))
-                    } else {
-                        Err(D::Error::custom("Failed to infer numeric type from it's value"))
+            }
+        }
+
+        impl<'de> Deserialize<'de> for NumberContent {
+            fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+            where
+                D: Deserializer<'de>,
+            {
+                let mut v: serde_json::Value = Value::deserialize(deserializer)?;
+                let as_object = v.as_object_mut().ok_or(D::Error::custom("Object value expected"))?;
+                match as_object.remove("subtype") {
+                    Some(subtype) => {
+                    $(
+                        if subtype == stringify!($ty) {
+                            if as_object.is_empty() {
+                                Ok(Self::$def())
+                            } else {
+                                let inner = number_content::$as::deserialize(v).map_err(D::Error::custom)?;
+                                Ok(NumberContent::$as(inner))
+                            }
+                        }
+                        else
+                    )*
+                        {
+                            //TODO: generate static array with variant names and pass it to
```

Comment:
TODO here

Reply:
An item left over from before this PR and got added to the diff as a result of the refactor. I think the idea was to make the serde error better, but we eventually worked around it. I will remove it.